File: ai-engine/nlp-processor/src/services/strategy_client.py
"""
Strategy Service Client

调用策略服务的 HTTP 客户端
"""

import logging
from typing import Any, Dict, List, Optional

import httpx

from ..config import get_settings

logger = logging.getLogger(__name__)


class StrategyServiceError(Exception):
    """策略服务错误"""

    def __init__(self, message: str, status_code: Optional[int] = None):
        super().__init__(message)
        self.status_code = status_code


class StrategyClient:
    """策略服务客户端"""

    def __init__(self, base_url: Optional[str] = None, timeout: float = 30.0):
        """
        初始化策略服务客户端

        Args:
            base_url: 策略服务基础 URL
            timeout: 请求超时时间（秒）
        """
        settings = get_settings()
        self.base_url = base_url or settings.strategy_service_url
        self.timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None

    async def _get_client(self) -> httpx.AsyncClient:
        """获取 HTTP 客户端"""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                base_url=self.base_url,
                timeout=self.timeout,
                headers={
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                },
            )
        return self._client

    async def close(self) -> None:
        """关闭客户端连接"""
        if self._client and not self._client.is_closed:
            await self._client.aclose()
            self._client = None

    async def create_strategy(
        self,
        name: str,
        symbol: str,
        strategy_type: str,
        config: Dict[str, Any],
        user_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        创建策略

        Args:
            name: 策略名称
            symbol: 交易对
            strategy_type: 策略类型
            config: 策略配置
            user_id: 用户 ID

        Returns:
            创建的策略数据（包含 strategy_id）
        """
        client = await self._get_client()

        payload = {
            "name": name,
            "symbol": symbol,
            "type": strategy_type,
            "config": config,
        }

        if user_id:
            payload["userId"] = user_id

        try:
            response = await client.post("/", json=payload)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"Strategy service error: {e.response.status_code} - {e.response.text}")
            raise StrategyServiceError(
                f"创建策略失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")

    async def get_strategy(self, strategy_id: str) -> Optional[Dict[str, Any]]:
        """
        获取策略详情

        Args:
            strategy_id: 策略 ID

        Returns:
            策略数据，不存在则返回 None
        """
        client = await self._get_client()

        try:
            response = await client.get(f"/{strategy_id}")
            if response.status_code == 404:
                return None
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"Get strategy error: {e.response.status_code}")
            raise StrategyServiceError(
                f"获取策略失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")

    async def list_strategies(
        self,
        user_id: Optional[str] = None,
        status: Optional[str] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        """
        获取策略列表

        Args:
            user_id: 用户 ID（可选，用于过滤）
            status: 策略状态（可选，用于过滤）
            limit: 返回数量限制
            offset: 偏移量

        Returns:
            策略列表
        """
        client = await self._get_client()

        params = {"limit": limit, "offset": offset}
        if user_id:
            params["userId"] = user_id
        if status:
            params["status"] = status

        try:
            response = await client.get("/", params=params)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"List strategies error: {e.response.status_code}")
            raise StrategyServiceError(
                f"获取策略列表失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")

    async def update_strategy(
        self,
        strategy_id: str,
        updates: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        更新策略

        Args:
            strategy_id: 策略 ID
            updates: 更新的字段

        Returns:
            更新后的策略数据
        """
        client = await self._get_client()

        try:
            response = await client.patch(f"/{strategy_id}", json=updates)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"Update strategy error: {e.response.status_code}")
            raise StrategyServiceError(
                f"更新策略失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")

    async def delete_strategy(self, strategy_id: str) -> bool:
        """
        删除策略

        Args:
            strategy_id: 策略 ID

        Returns:
            是否删除成功
        """
        client = await self._get_client()

        try:
            response = await client.delete(f"/{strategy_id}")
            return response.status_code in [200, 204]
        except httpx.HTTPStatusError as e:
            logger.error(f"Delete strategy error: {e.response.status_code}")
            raise StrategyServiceError(
                f"删除策略失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")

    async def start_strategy(self, strategy_id: str) -> Dict[str, Any]:
        """
        启动策略

        Args:
            strategy_id: 策略 ID

        Returns:
            启动结果
        """
        client = await self._get_client()

        try:
            response = await client.post(f"/{strategy_id}/start")
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"Start strategy error: {e.response.status_code}")
            raise StrategyServiceError(
                f"启动策略失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")

    async def stop_strategy(self, strategy_id: str) -> Dict[str, Any]:
        """
        停止策略

        Args:
            strategy_id: 策略 ID

        Returns:
            停止结果
        """
        client = await self._get_client()

        try:
            response = await client.post(f"/{strategy_id}/stop")
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            logger.error(f"Stop strategy error: {e.response.status_code}")
            raise StrategyServiceError(
                f"停止策略失败: {e.response.text}",
                status_code=e.response.status_code,
            )
        except httpx.RequestError as e:
            logger.error(f"Strategy service request error: {e}")
            raise StrategyServiceError(f"策略服务连接失败: {str(e)}")


# =============================================================================
# Singleton Instance
# =============================================================================

_client: Optional[StrategyClient] = None


def get_strategy_client() -> StrategyClient:
    """获取策略服务客户端单例"""
    global _client
    if _client is None:
        _client = StrategyClient()
    return _client


async def close_strategy_client() -> None:
    """关闭策略服务客户端"""
    global _client
    if _client is not None:
        await _client.close()
        _client = None

File: backend/ai-orchestrator/src/services/strategy-client.ts
/**
 * Strategy Service 客户端
 *
 * 与 Strategy Service 通信，实现策略的 CRUD 操作
 */

// =============================================================================
// 配置
// =============================================================================

const STRATEGY_SERVICE_URL = process.env.STRATEGY_SERVICE_URL || 'http://localhost:4002'

// =============================================================================
// 类型定义
// =============================================================================

export interface Factor {
  id: string
  name?: string
  params?: Record<string, unknown>
}

export interface RiskManagement {
  stopLoss?: number
  takeProfit?: number
  positionSize?: number
  maxDrawdown?: number
  dailyLossLimit?: number
}

export interface StrategyConfig {
  name: string
  symbol: string
  timeframe?: string
  factors: Factor[]
  entryConditions: string[]
  exitConditions?: string[]
  riskManagement?: RiskManagement
  description?: string
  tags?: string[]
}

export interface Strategy {
  id: string
  userId: string
  name: string
  symbol: string
  timeframe?: string
  factors: Factor[]
  entryConditions: string[]
  exitConditions?: string[]
  riskManagement?: RiskManagement
  status: 'draft' | 'active' | 'paused' | 'stopped' | 'archived'
  version: number
  description?: string
  tags?: string[]
  performance?: StrategyPerformance
  createdAt: string
  updatedAt: string
}

export interface StrategyPerformance {
  totalPnL: number
  totalPnLPercent: number
  winRate: number
  totalTrades: number
  sharpeRatio?: number
  maxDrawdown?: number
  lastUpdated: string
}

export interface CreateStrategyResponse {
  success: boolean
  strategy?: Strategy
  strategyId?: string
  error?: string
}

export interface UpdateStrategyResponse {
  success: boolean
  strategy?: Strategy
  newVersion?: number
  error?: string
}

export interface ListStrategiesResponse {
  success: boolean
  strategies: Strategy[]
  total: number
  page: number
  pageSize: number
}

export interface StrategyActionResponse {
  success: boolean
  strategyId: string
  status: string
  message?: string
  error?: string
}

// =============================================================================
// Strategy Service 客户端
// =============================================================================

export class StrategyClientService {
  private baseUrl: string
  private isAvailable: boolean | null = null

  constructor() {
    this.baseUrl = STRATEGY_SERVICE_URL
  }

  /**
   * 健康检查
   */
  async healthCheck(): Promise<{ healthy: boolean; latency: number }> {
    const start = Date.now()
    try {
      const response = await fetch(`${this.baseUrl}/health`, {
        method: 'GET',
        signal: AbortSignal.timeout(5000),
      })

      this.isAvailable = response.ok
      return {
        healthy: response.ok,
        latency: Date.now() - start,
      }
    } catch {
      this.isAvailable = false
      return {
        healthy: false,
        latency: Date.now() - start,
      }
    }
  }

  /**
   * 创建策略
   */
  async createStrategy(
    config: StrategyConfig,
    userId: string
  ): Promise<CreateStrategyResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/strategies`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-User-Id': userId,
        },
        body: JSON.stringify(config),
        signal: AbortSignal.timeout(30000),
      })

      if (!response.ok) {
        const error = await response.json().catch(() => ({})) as { message?: string }
        return {
          success: false,
          error: error.message || `Failed to create strategy: ${response.status}`,
        }
      }

      const data = await response.json() as Strategy
      return {
        success: true,
        strategy: data,
        strategyId: data.id,
      }
    } catch (error) {
      console.error('[StrategyClient] Create strategy failed:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to create strategy',
      }
    }
  }

  /**
   * 获取策略详情
   */
  async getStrategy(strategyId: string, userId: string): Promise<Strategy | null> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/strategies/${strategyId}`, {
        method: 'GET',
        headers: {
          'X-User-Id': userId,
        },
        signal: AbortSignal.timeout(10000),
      })

      if (!response.ok) {
        return null
      }

      return await response.json() as Strategy
    } catch (error) {
      console.error('[StrategyClient] Get strategy failed:', error)
      return null
    }
  }

  /**
   * 列出用户策略
   */
  async listStrategies(
    userId: string,
    options?: {
      status?: 'all' | 'active' | 'paused' | 'draft'
      limit?: number
      page?: number
      sortBy?: 'createdAt' | 'updatedAt' | 'performance'
    }
  ): Promise<ListStrategiesResponse> {
    try {
      const params = new URLSearchParams()
      if (options?.status && options.status !== 'all') {
        params.set('status', options.status)
      }
      if (options?.limit) {
        params.set('limit', String(options.limit))
      }
      if (options?.page) {
        params.set('page', String(options.page))
      }
      if (options?.sortBy) {
        params.set('sortBy', options.sortBy)
      }

      const url = `${this.baseUrl}/api/v1/strategies?${params.toString()}`
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'X-User-Id': userId,
        },
        signal: AbortSignal.timeout(15000),
      })

      if (!response.ok) {
        return {
          success: false,
          strategies: [],
          total: 0,
          page: options?.page || 1,
          pageSize: options?.limit || 20,
        }
      }

      const data = await response.json() as {
        strategies: Strategy[]
        total: number
        page: number
        pageSize: number
      }

      return {
        success: true,
        ...data,
      }
    } catch (error) {
      console.error('[StrategyClient] List strategies failed:', error)
      return {
        success: false,
        strategies: [],
        total: 0,
        page: options?.page || 1,
        pageSize: options?.limit || 20,
      }
    }
  }

  /**
   * 更新策略参数
   */
  async updateStrategy(
    strategyId: string,
    userId: string,
    updates: Partial<StrategyConfig>
  ): Promise<UpdateStrategyResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/strategies/${strategyId}`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
          'X-User-Id': userId,
        },
        body: JSON.stringify(updates),
        signal: AbortSignal.timeout(15000),
      })

      if (!response.ok) {
        const error = await response.json().catch(() => ({})) as { message?: string }
        return {
          success: false,
          error: error.message || `Failed to update strategy: ${response.status}`,
        }
      }

      const data = await response.json() as Strategy
      return {
        success: true,
        strategy: data,
        newVersion: data.version,
      }
    } catch (error) {
      console.error('[StrategyClient] Update strategy failed:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to update strategy',
      }
    }
  }

  /**
   * 启动策略
   */
  async startStrategy(strategyId: string, userId: string): Promise<StrategyActionResponse> {
    return this.performStrategyAction(strategyId, userId, 'start')
  }

  /**
   * 暂停策略
   */
  async pauseStrategy(strategyId: string, userId: string): Promise<StrategyActionResponse> {
    return this.performStrategyAction(strategyId, userId, 'pause')
  }

  /**
   * 停止策略
   */
  async stopStrategy(strategyId: string, userId: string): Promise<StrategyActionResponse> {
    return this.performStrategyAction(strategyId, userId, 'stop')
  }

  /**
   * 删除策略
   */
  async deleteStrategy(strategyId: string, userId: string): Promise<StrategyActionResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/strategies/${strategyId}`, {
        method: 'DELETE',
        headers: {
          'X-User-Id': userId,
        },
        signal: AbortSignal.timeout(10000),
      })

      if (!response.ok) {
        const error = await response.json().catch(() => ({})) as { message?: string }
        return {
          success: false,
          strategyId,
          status: 'error',
          error: error.message || `Failed to delete strategy: ${response.status}`,
        }
      }

      return {
        success: true,
        strategyId,
        status: 'deleted',
        message: 'Strategy deleted successfully',
      }
    } catch (error) {
      console.error('[StrategyClient] Delete strategy failed:', error)
      return {
        success: false,
        strategyId,
        status: 'error',
        error: error instanceof Error ? error.message : 'Failed to delete strategy',
      }
    }
  }

  /**
   * 编译策略 (生成 RuntimePlan)
   */
  async compileStrategy(
    strategyId: string,
    userId: string,
    version?: number
  ): Promise<{ success: boolean; runtimePlanId?: string; hash?: string; error?: string }> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/strategies/${strategyId}/compile`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-User-Id': userId,
        },
        body: JSON.stringify({ version }),
        signal: AbortSignal.timeout(30000),
      })

      if (!response.ok) {
        const error = await response.json().catch(() => ({})) as { message?: string }
        return {
          success: false,
          error: error.message || `Failed to compile strategy: ${response.status}`,
        }
      }

      const data = await response.json() as { runtimePlanId: string; hash: string }
      return {
        success: true,
        runtimePlanId: data.runtimePlanId,
        hash: data.hash,
      }
    } catch (error) {
      console.error('[StrategyClient] Compile strategy failed:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Failed to compile strategy',
      }
    }
  }

  /**
   * 执行策略操作 (start/pause/stop)
   */
  private async performStrategyAction(
    strategyId: string,
    userId: string,
    action: 'start' | 'pause' | 'stop'
  ): Promise<StrategyActionResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/api/v1/strategies/${strategyId}/${action}`, {
        method: 'POST',
        headers: {
          'X-User-Id': userId,
        },
        signal: AbortSignal.timeout(15000),
      })

      if (!response.ok) {
        const error = await response.json().catch(() => ({})) as { message?: string }
        return {
          success: false,
          strategyId,
          status: 'error',
          error: error.message || `Failed to ${action} strategy: ${response.status}`,
        }
      }

      const data = await response.json() as { status: string; message?: string }
      return {
        success: true,
        strategyId,
        status: data.status,
        message: data.message,
      }
    } catch (error) {
      console.error(`[StrategyClient] ${action} strategy failed:`, error)
      return {
        success: false,
        strategyId,
        status: 'error',
        error: error instanceof Error ? error.message : `Failed to ${action} strategy`,
      }
    }
  }

  /**
   * 检查服务是否可用（使用缓存）
   */
  async isServiceAvailable(): Promise<boolean> {
    if (this.isAvailable === null) {
      const health = await this.healthCheck()
      this.isAvailable = health.healthy
    }
    return this.isAvailable
  }

  /**
   * 重置可用性缓存
   */
  resetAvailabilityCache(): void {
    this.isAvailable = null
  }
}

// 单例导出
export const strategyClient = new StrategyClientService()
